### ISSUE TITLE
Rendering OOM creates flashes on Vulkan due to bug in render recovery

### ISSUE THREAD

**Issue by DGriffin91** [bevyengine/bevy#23183]
## Bevy main 2494a37573c03ed011a11a373d0f27b1eb8ed450 

About half the scene flashes in and out of existence every other frame. 

This appears to only happen on the vulkan backend (on both linux and windows). If I set `WGPU_BACKEND=dx12` on windows I am able to open multiple instances of bistro without issues, and can see that I am spilling into RAM.

On vulkan the issue seems to appear around the time I run out of VRAM. When loading, it usually (but not always) hits 16GB, starts flashing, with the VRAM usage settling down to 11GB. Every once in a while (rarely), the usage never hits 16GB and the scene loads correctly on main. 

If `--no-mip-generation` is set, flashing doesn't occur on my pc. But if I open 3 instances of bistro at the same time, flashing still occurs. On the commit before bevy updated to wgpu 0.28 I am able to simultaneously open 3 instances of bistro *with* mip generation and no compression and after taking a while the scenes seem to all load correctly with no flashing.

Tried bisecting, ran into different issues/crashes at different times. There may be UB, system ordering ambiguity, or race conditions involved, as the crashing doesn't seem to be 100% consistent.

### Timelime
Before 543b305adce77f84cffaa964b631724ea7051135 the scene loads as expected. Each of these are the first commit that the behavior changes:

- 1/22: 543b305adce77f84cffaa964b631724ea7051135 crashes with `Trying to destroy a SwapchainAcquireSemaphore that is still in use by a SurfaceTexture`
- 1/25: d734e7feb060900f4a79f5611126c272e271e31e crashes with `wgpu error: Out of Memory`
- 2/5: 71dd9ea7db45b9d32f38d08c8abd45e98b1fdc31 doesn't crash but scene is flashing

Tested on Linux & Windows, RTX4070ti (16GB), 7950x with bistro example in `large_scenes`

Tried wgpu 28.0.1, it had the same flashing issue.

**Comment by alice-i-cecile** [comment]
Might be related to #23143, which is reporting similar flashes.

**Comment by mockersf** [comment]
can't reproduce on macOS on main

**Comment by alice-i-cecile** [comment]
[Messages on Discord](https://discord.com/channels/691052431525675048/743663924229963868/1477947269428023347) indicate that this is likely a VRAM OOM issue caused by increases in GPU memory usage driven by the wgpu upgrade. The flashing is likely due to mip map textures being unloaded and reloaded.

**Comment by DGriffin91** [comment]
Based on François's suggestion I tried adding:
```rs
app.insert_resource(bevy::render::error_handler::RenderErrorHandler(
        |_, _, _| bevy::render::error_handler::RenderErrorPolicy::StopRendering,
))
```

With this it flickers once then locks up without crashing and starts leaking lots of memory (ram) presumably until it hits the *other* OOM. Also it not spamming the console let me see this error:
`ERROR bevy_render::error_handler: Caught rendering error: Out of Memory`
Which I had seen previously before the PR that started the version of this were if flickers.

There's also other errors like:
`Validation Error Caused by: In Queue::write_texture Texture with '' label is invalid`
`Validation Error Caused by: In Queue::submit In a set_bind_group command BindGroup with 'StandardMaterial' label is invalid`
But once it locks up it doesn't spam those it prints a bunch and then stops printing anything new.

**Comment by alice-i-cecile** [comment]
I agree with @JMS55 that we should cut this from the milestone: this is fundamentally a wgpu problem, and not critical.

**Comment by DGriffin91** [comment]
> I agree with [@JMS55](https://github.com/JMS55) that we should cut this from the milestone: this is fundamentally a wgpu problem, and not critical.

@alice-i-cecile I strongly believe the current failure mode of flashing is a critical safety issue for people who are sensitive.

**Comment by alice-i-cecile** [comment]
If you have suggestions about how to fix this, I'm all ears. I would love to make this better.

**Comment by DGriffin91** [comment]
@alice-i-cecile I believe the flashing is due to the new render recovery (https://github.com/bevyengine/bevy/commit/71dd9ea7db45b9d32f38d08c8abd45e98b1fdc31) `RenderErrorHandler` not working correctly. (See https://github.com/bevyengine/bevy/issues/23183#issuecomment-3989640429) 

In regard to handling errors, there a comment that says: 
> This is what we've always done historically

https://github.com/bevyengine/bevy/blob/main/crates/bevy_render/src/error_handler.rs#L61-L67

But previously OOM and similar errors would result in a crash, rather than result in flashing. The OOM highlighted the issue, but is not the cause of flashing in particular. Note that I have a GPU with 16GB of VRAM. It's going to be much easier for others with less VRAM to hit this particular issue. And even if the OOM is fixed in wgpu there could be other things that now in bevy 0.19 trigger flashing that previously would crash.

**Comment by alice-i-cecile** [comment]
Ah, that's much clearer. cc @atlv24 :)

**Comment by kfc35** [comment]
Would just updating the default `RenderErrorHandler` to `StopRendering` when it encounters an `ErrorType::OutOfMemory` suffice?

 Or should we maybe have an additional `RenderErrorPolicy` that implements exiting the app, and have that be called upon `OutOfMemory`?

**Comment by DGriffin91** [comment]
If I comment this out it crashes like it did prior to the [render recovery PR](https://github.com/bevyengine/bevy/pull/22761): 
https://github.com/bevyengine/bevy/blob/ce2bf99b6a001161924770a7855f4024b67138eb/crates/bevy_render/src/error_handler.rs#L109-L119

**Comment by DGriffin91** [comment]
> Would just updating the default `RenderErrorHandler` to `StopRendering` when it encounters an `ErrorType::OutOfMemory` suffice?
> 
> Or should we maybe have an additional `RenderErrorPolicy` that implements exiting the app, and have that be called upon `OutOfMemory`?

@kfc35 We can't have flash strobing people be a failure mode just in general because of the safety issue. So I think it's really important that we don't just special case the OOM and instead fix the underlying issue. The OOM is basically just a canary.
iirc the new view binding layout issue was also resulting in strobing in one case, but it would take more looking into to get something easily reproducible

**Comment by kfc35** [comment]
RE: the new view binding layout also causing strobing, let’s keep that conversation in #24084 so as not to distract from this one. I’ve also found strobing since that merge and am slowly combing through the problems

**Comment by kfc35** [comment]
> If I comment this out it crashes like it did prior to the [render recovery PR](https://github.com/bevyengine/bevy/pull/22761):

This makes sense to me, since according to https://github.com/gfx-rs/wgpu/blob/bb83a5824a5c84f315c62368eac1bdaf69d82b90/wgpu/src/api/device.rs#L787-L793 the default is to panic unless a handler has been defined. The handler, as written, passes it off to eventually become converted into a `RenderError`. The default error handler just puts the render into a `RenderState::Ready` state, and then the render just tries to start as normal. Then there’s just a loop of OOM, ready, and then trying to run the same render schedule without modification. I wouldn’t be surprised if that’s the source of the strobing.

To be most prohibitive strobing, I would suggest that we handle *all* uncaptured errors (OOM, Validation, Internal) and the singled out device error to stop the app. I’ve seen my fair share of strobing just due to constant Validation errors that are ignored, so I’m pretty sure those should just quit the app as well. I haven’t seen an internal error before, and I don’t know much about device lost but I assume it would be the same deal. I agree with you that we should first and foremost prevent this from becoming a safety issue, and then we can look into retrying for any recoverable errors more conservatively and as we see fit.

It doesn’t necessarily have to panic (can you stop an app with just access to the main world? Just grab the message writer for app exit and write an app exit?), but if we really want to go with the default of what wgpu does, it can just panic.
